### PR TITLE
Fix #65: verify certificate chain

### LINKED ISSUE #65: Verify certificate chain
* Each cert should be valid within current date (before / after)
* Chain of trust (root fingerprint should match root_hast constant)
* Check attributes, subject, key type, key usage

See [python implementation](https://github.com/mozilla-services/python-autograph-utils/blob/eff5244da503546654285e12f7a643a24c4fbb59/autograph_utils/__init__.py#L345-L386)

### PR DESCRIPTION
**Pull request by leplatrem**
The current code only verifies the signature of the data with the leaf certificate. But the chain of certificate should also be verified

* [x] Working implementation with `ring`
* [ ] Add `rc_crypto` implementation
* [x] Add ability to mock time for expiration tests
* [x] Add unit test with whole cert chain

### CODE REVIEW

File: src/client/signatures/x509.rs [hunk 1]
```diff
@@ -14,20 +14,32 @@ pub enum X509Error {
...
+
+    let mut x509s = vec![];
+    for block in blocks {
+        let (_, pem) = x509_parser::pem::parse_x509_pem(&block)?;
+
+        if pem.label != "CERTIFICATE" {
+            return Err(X509Error::WrongPEMType(pem.label));
+        }
+
+        let (_, x509) = x509_parser::parse_x509_certificate(&pem.contents)?;
```
Comment (leplatrem) [inline src/client/signatures/x509.rs:31]:
```
error[E0597]: `pem.contents` does not live long enough
  --> src/client/signatures/x509.rs:31:61
   |
31 |         let (_, x509) = x509_parser::parse_x509_certificate(&pem.contents)?;
   |                                                             ^^^^^^^^^^^^^ borrowed value does not live long enough
...
34 |     }
   |     - `pem.contents` dropped here while still borrowed
35 | 
36 |     for cert in x509s.iter() {
   |                 ----- borrow later used here

error[E0597]: `x509` does not live long enough
```

File: src/client/signatures/rc_crypto_verifier.rs [hunk 2]
```diff
@@ -12,13 +12,29 @@ pub struct RcCryptoVerifier {}
...
-        let public_key_bytes = x509::extract_public_key(pem_bytes)?;
+        let pems = x509::parse_certificate_chain(&pem_bytes)?;
+
+        let certs: Vec<x509::X509Certificate> = pems
+            .iter()
+            .map(|pem| match x509::parse_x509_certificate(&pem) {
+                Ok(cert) => Ok(cert),
+                Err(e) => Err(e),
+            })
+            .collect::<Result<Vec<x509::X509Certificate>, _>>()?;
```
Comment (rfk) [inline src/client/signatures/rc_crypto_verifier.rs:29]: I guess we need to copy (or re-use) the `// Verify signature chain` logic from the `ring` implementation into this one as well.

Without thinking too deeply about it, would it make sense to factor that out into a shared default helper method on the Trait? Along the lines of:

```
pub trait Verification {
    // ... existing default methods here ...

    fn verify_certificate_chain(&self, certs: Vec<x509::X509Certificate>, root_hash: &str) -> Result<x509::X509Certificate, SignatureError> {
        // Do all the parsing and stuff.
        // Call methods on `self` to delegate to verification of each signature to the trait implemetor
        // Return the leaf certificate by use by calling code.
    }
```

It seems like *most* of the code there is parsing stuff that could usefully be shared between implementations.

**Review by rfk — COMMENTED** [review]
Thanks @leplatrem! What's going on here at a code level makes sense to me, and I left a few comments below.

Zooming out a level, the amount of certificate-parsing code that's happening here makes me a little wary. I don't have a specific concern or request right now, but want to share a bit of perspective from previous conversations we had around shipping Rust code in Fenix w.r.t networking and crypto.

The genesis of `viaduct` and `rc_crypto` was essentially that Mozilla has invested in a particular quality and policy bar for our networking and crypto code, and we want to make sure our products use that uniformly where possible. So for example, the Rust code in Fenix was not allowed to use the default `reqwest` module's networking stack, because that comes with a third-party TLS setup, and Mozilla hasn't audited either the code or the policy of that stack, so we might be unexpected exposing our users to risks. We instead arranged for the Rust code to make network requests by piping them through to Necko, the same networking layer responsible for all requests in the browser.

Note that the concern here wasn't just about the integrity of particular crypto routines, it was about maintaining the holistic security and policy stance offered by Necko. (What algorithms are allowed? What roots do we trust? Are some certificates pinned? Should we use DoH? etc).

I'm concerned that we may be headed towards a similar (but smaller!) situation here with the parsing of certificates and handling of certificate chains. NSS seems likely to have its own routines for parsing x509 certificates, perhaps even for verifying chains of them like we're doing here. If it does, then the right thing for our users is probably to expose those higher-level functions from NSS and call them directly, rather than re-implement them on top of NSS's lower-level primitives.

Alternately, we should make sure we have explicit acknowledgement from the security/policy folks that they're OK with the architecture here.

Who is currently on the hook to review this setup from a security/policy perspective?

(I don't think that broader question should block us from working on this in the short term, but I think we should proactively try to avoid a situation months from now where someone like :ekr looks at what we're doing here and says "why aren't you using NSS for this?").

**Comment by rfk** [comment]
A concrete example of what I mean: the desktop Remote Settings client uses [`nsIContentSignatureVerifier`](https://searchfox.org/mozilla-central/source/security/manager/ssl/ContentSignatureVerifier.cpp), which relies on [`BuildCertChain`](https://searchfox.org/mozilla-central/source/security/manager/ssl/ContentSignatureVerifier.cpp), which AFAICT comes from [`mozilla::pkix::BuildCertChain`](https://searchfox.org/mozilla-central/rev/bb37e6fe8bbe383a57a4ad21a201e26416613ca1/security/nss/lib/mozpkix/lib/pkixbuild.cpp#386) in C++. I don't know how closely this maps to the [`PKIX_BuildChain`](https://searchfox.org/mozilla-central/rev/bb37e6fe8bbe383a57a4ad21a201e26416613ca1/security/nss/lib/libpkix/include/pkix.h#240) C function exported by libpkix, but should we be trying to expose that to Rust and call it directly rather than doing this lower-level certificate walking?

(I genuinely don't know how tractable that would be or how much value it would provide relative to the approach taken here, just trying to get my head around the problem space).

File: src/client/signatures/ring_verifier.rs [hunk 3]
```diff
@@ -5,20 +5,95 @@
...
+        // Check certificate validity.
+        // TODO: take clock skew into account
+        // TODO: mock from tests
+        let now = ASN1Time::now();
+        if !leaf_cert.tbs_certificate.validity.is_valid_at(now) {
+            return Err(SignatureError::CertificateExpired);
+        }
+
+        // Verify signature chain.
+        for pair in certs.windows(2) {
```
[the remark this answers is not in the document]
Comment (leplatrem) [reply src/client/signatures/ring_verifier.rs:64]: Greg replied:

> Autograph has a test in the monitor for out of order chains https://github.com/mozilla-services/autograph/blob/main/tools/autograph-monitor/contentsignaturepki_test.go#L553, and verifies chains before uploading them https://github.com/mozilla-services/autograph/blob/main/signer/contentsignaturepki/x509.go#L68So it's a good thing to have a test for, but should never happen.
> I wouldn't try to test all permutations of invalid chains for example, but it is good to verify that each link of the chain is signed by its issuer.

**Comment by leplatrem** [comment]
Thanks @rfk for your feedback here! I created #98 to address this precise point. 

Ideally, I'd like to keep the two implementations in the repo (ie. pure Rust with `ring`+`x509-parser`, and `rc_crypto`+`libpkix`). Let's how I can refactor that better.

> Who is currently on the hook to review this setup from a security/policy perspective?

No one yet. I will try to figure this out.

**Comment by rfk** [comment]
> Ideally, I'd like to keep the two implementations in the repo (ie. pure Rust with ring+x509-parser, and rc_crypto+libpkix).
> Let's how I can refactor that better.

Agreed, yes, I think that makes a lot of sense, especially if we can use cargo features to disable them at build time for integration in e.g. Fenix or one day in mozilla-central.

File: src/client/signatures/ring_verifier.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (rfk) [reply src/client/signatures/ring_verifier.rs:64]: > good to verify that each link of the chain is signed by its issuer

If by "good" you mean "crucial to the security of the system", sure... 😅 

FWIW I'm very supporting of having this as a requirement that the server always presents certificates in the correct order 👍

**Comment by leplatrem** [comment]
Ok, I refactored the trait code. I'm not entirely convinced by the naming and granularity that I chose, I will see how it looks if I follow your suggestion.

In the mean time I also realized that `rc_crypto` doesn't expose the RSA primitives, necessary to verify the intermediate certificates.

**Comment by leplatrem** [comment]
I think it would make sense the review and merge this PR, even though the rc_crypto does not match our mid-term requirements.

The current state of this PR:
- the `ring` implementation is complete
- the trait is prepared for the future drop-in replacement of NSS (based [on option D in decision brief](https://docs.google.com/document/d/1MOyx9sK_NWo6c9STN5O6P0dTCfIBc0ZOjhd7fZPtjuk/edit#))
- the rc_crypto code still uses the external x509-parser **just for the unit tests to pass**

What do you think?

**Comment by leplatrem** [comment]
I merged this because:
- I'm happy with the current state
- I need a clean main branch and tag to start the security review process
- the client isn't used in production
- if any review is given now, I can still open a follow-up PR

**Pull request merged**